{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Gonatopoulos-Pournatzis et al (2020) double perturbation screens (CHyMErA screens)\n",
    "\n",
    "**Genetic interaction mapping and exon-resolution functional genomics with a hybrid Cas9–Cas12a platform**  \n",
    "https://www.nature.com/articles/s41587-020-0437-z  \n",
    "\n",
    "Notes:\n",
    "- Target 1,344 genes that represent paralog pairs\n",
    "- HAP1 and RPE1 cells\n",
    "- Effect size represents mean difference in expected and observed LFC at early (T12 for HAP1, T18 for RPE1) or late (T18 for HAP1, T24 for RPE1) time points\n",
    "\n",
    "**Input**: Supplemental Table 8 - contains several sheets of interest: Summary, HAP1.T12, HAP1.T18, RPE1.T18, RPE1.T24\n",
    "\n",
    "**Output**: Paralog pairs from screen annotated with negative GI hits and observed LFC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import scipy.stats as stats\n",
    "from matplotlib_venn import venn2\n",
    "import re\n",
    "import os\n",
    "\n",
    "get_data_path = lambda folders, fname: os.path.normpath(os.environ['3RD_PARTY_DIR']+'/'+'/'.join(folders) +'/'+ fname)\n",
    "get_local_data_path = lambda folders, fname: os.path.normpath('../../local_data/' +'/'.join(folders) +'/'+ fname)\n",
    "\n",
    "file_data = get_data_path(['GI_screens', 'gonatopoulus_2020'], 'Table_S8.xlsx')\n",
    "file_all_pairs = lambda ens_v: get_local_data_path(['processed', 'ensembl'+ens_v], 'all_pairs.csv')\n",
    "\n",
    "# Output\n",
    "file_chymera_pairs_orig = get_local_data_path(['processed','screen_pairs'], 'chymera_pairs_orig.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Process Table S8 from paper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "N pairs: 688\n",
      "N w/ GI: 219 122\n",
      "N w/ GI in either: 290\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene1</th>\n",
       "      <th>gene2</th>\n",
       "      <th>hap1_gi_type</th>\n",
       "      <th>rpe1_gi_type</th>\n",
       "      <th>hap1_early_effect_size</th>\n",
       "      <th>hap1_late_effect_size</th>\n",
       "      <th>rpe1_early_effect_size</th>\n",
       "      <th>rpe1_late_effect_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SEC23A</td>\n",
       "      <td>SEC23B</td>\n",
       "      <td>shared</td>\n",
       "      <td>shared</td>\n",
       "      <td>-3.028953</td>\n",
       "      <td>-3.424526</td>\n",
       "      <td>-2.306238</td>\n",
       "      <td>-2.103825</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CUL4A</td>\n",
       "      <td>CUL4B</td>\n",
       "      <td>shared</td>\n",
       "      <td>shared</td>\n",
       "      <td>-2.842330</td>\n",
       "      <td>-3.439328</td>\n",
       "      <td>-1.826966</td>\n",
       "      <td>-1.543943</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    gene1   gene2 hap1_gi_type rpe1_gi_type  hap1_early_effect_size  \\\n",
       "0  SEC23A  SEC23B       shared       shared               -3.028953   \n",
       "1   CUL4A   CUL4B       shared       shared               -2.842330   \n",
       "\n",
       "   hap1_late_effect_size  rpe1_early_effect_size  rpe1_late_effect_size  \n",
       "0              -3.424526               -2.306238              -2.103825  \n",
       "1              -3.439328               -1.826966              -1.543943  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary = pd.read_excel(file_data, sheet_name=\"Summary\", skiprows=1)\n",
    "summary = summary.drop(columns=['torin_gi_type', 'torin_early_effect_size', 'torin_late_effect_size', 'HAP1 +/- Torin'])\n",
    "print('N pairs:', summary.shape[0])\n",
    "print('N w/ GI:', summary[summary.hap1_gi_type!='none'].shape[0], summary[summary.rpe1_gi_type!='none'].shape[0])\n",
    "print('N w/ GI in either:', summary[(summary.hap1_gi_type!='none')|(summary.rpe1_gi_type!='none')].shape[0])\n",
    "summary[:2]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Assign original hit calls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Num HAP1 SL: 214 / 688\n",
      "Num RPE1 SL: 107 / 688\n",
      "SL in either cell line: 271 = 39.39%\n",
      "SL in both cell lines: 50 = 7.27%\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "      <th>hap1_hit</th>\n",
       "      <th>rpe1_hit</th>\n",
       "      <th>n_hits</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SEC23A</td>\n",
       "      <td>SEC23B</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       A1      A2  hap1_hit  rpe1_hit  n_hits\n",
       "0  SEC23A  SEC23B      True      True       2"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Call negative GI if the effect is seen at either time point (shared), and effect size is negative at both time points\n",
    "screen_hits = summary.assign(\n",
    "    hap1_hit = summary.apply(lambda x: ((x.hap1_gi_type!='none') \n",
    "                                        and (x.hap1_late_effect_size < 0) and (x.hap1_early_effect_size < 0)), axis=1),\n",
    "    rpe1_hit = summary.apply(lambda x: ((x.rpe1_gi_type!='none')\n",
    "                                        and (x.rpe1_late_effect_size < 0) and (x.rpe1_early_effect_size < 0)), axis=1))\n",
    "\n",
    "print('Num HAP1 SL:', sum(screen_hits.hap1_hit), '/', screen_hits.shape[0])\n",
    "print('Num RPE1 SL:', sum(screen_hits.rpe1_hit), '/', screen_hits.shape[0])\n",
    "print('SL in either cell line: %d = %.2f%%' % (sum(screen_hits.hap1_hit | screen_hits.rpe1_hit),\n",
    "                                               sum(screen_hits.hap1_hit | screen_hits.rpe1_hit)/screen_hits.shape[0]*100))\n",
    "print('SL in both cell lines: %d = %.2f%%' % (sum(screen_hits.hap1_hit & screen_hits.rpe1_hit),\n",
    "                                              sum(screen_hits.hap1_hit & screen_hits.rpe1_hit)/screen_hits.shape[0]*100))\n",
    "\n",
    "# Count num cell lines in which pairs are SL\n",
    "screen_hits['n_hits'] = screen_hits[['hap1_hit', 'rpe1_hit']].sum(axis=1)\n",
    "screen_hits = screen_hits[['gene1','gene2','hap1_hit','rpe1_hit','n_hits']].rename(columns={'gene1':'A1','gene2':'A2'})\n",
    "screen_hits = screen_hits.rename(columns={'gene1':'A1','gene2':'A2'})\n",
    "screen_hits[:1]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Merge with ensembl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "N= 73296\n",
      "N= 179968\n"
     ]
    }
   ],
   "source": [
    "all_pairs_93 = pd.read_csv(file_all_pairs('93'))[['A1','A2']]\n",
    "print('N=', all_pairs_93.shape[0])\n",
    "all_pairs_102 = pd.read_csv(file_all_pairs('102'))[['A1','A2']]\n",
    "print('N=', all_pairs_102.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overlap w/ Ens.93: 658\n",
      "Overlap w/ Ens.102: 642\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sorted_gene_pair</th>\n",
       "      <th>hap1_hit</th>\n",
       "      <th>rpe1_hit</th>\n",
       "      <th>n_hits</th>\n",
       "      <th>in_93</th>\n",
       "      <th>in_102</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SEC23A_SEC23B</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CUL4A_CUL4B</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sorted_gene_pair  hap1_hit  rpe1_hit  n_hits  in_93  in_102\n",
       "0    SEC23A_SEC23B      True      True       2   True    True\n",
       "1      CUL4A_CUL4B      True      True       2   True    True"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Merge with our Ensembl list of paralog pairs\n",
    "screen_pairs = pd.merge(screen_hits, all_pairs_93, how='left', indicator='in_93')\n",
    "screen_pairs = pd.merge(screen_pairs, all_pairs_102, how='left', indicator='in_102')\n",
    "screen_pairs['in_93'] = screen_pairs['in_93'] == 'both'\n",
    "screen_pairs['in_102'] = screen_pairs['in_102'] == 'both'\n",
    "assert(screen_hits.shape[0] == screen_pairs.shape[0])\n",
    "print('Overlap w/ Ens.93:', sum(screen_pairs.in_93))\n",
    "print('Overlap w/ Ens.102:', sum(screen_pairs.in_102))\n",
    "assert(screen_hits.shape[0] == screen_pairs.shape[0])\n",
    "\n",
    "# Sorted gene pair column\n",
    "screen_pairs.insert(0, 'sorted_gene_pair', screen_pairs.apply(lambda x: '_'.join(sorted([x.A1, x.A2])), axis=1))\n",
    "screen_pairs = screen_pairs.drop(columns=['A1','A2'])\n",
    "screen_pairs[:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAALUAAACVCAYAAAAe/7ufAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8vihELAAAACXBIWXMAAAsTAAALEwEAmpwYAAAUKUlEQVR4nO2de3iU1Z3HP2cSQhIgCTAQCJIEEAjIbSyoKIhooUqr1kvtVt3qPm29dfex213akXYrrS0M1Ue77fZi19Ze9qK17WrxgopaBUXB8ga5Ry7hFkLyArlnJpmZs3+8EwghATJ575zP87wPyTuZc34zfOc3v3PO7/yOkFKiUPiJgNMGKBRmo0St8B1K1ArfoUSt8B1K1ArfoUSt8B1K1ArfoUSt8B1K1ArfoUSt8B1K1ArfoUSt8B1K1ArfoUSt8B1K1ArfoUSt8B2ZThvgFiKang3kpq5sjPemX+pfAUQ7Xa0dP4dDQbXLwmWI83HnS0TTC4DCTlc+6X9rxYBmQAdqU9fRcCiY6LulinTwvagjmp4BjOCkgIcD/S3uNgEcAQ6lrlrl0e3Dl6KOaLoARgLjgTFAlrMWEQV2AtvDoWCDw7b4Hl+JOqLpucAkYCIw0GFzeqIK2A7sDYeCSaeN8SO+EHVE00cAU4BSvDOjEwUqMLx3vdPG+AlPizo14JsNjHbYlL6yB1ivQhNz8KSoI5reH/gEMBnveOazkQS2AX8Lh4Ixp43xMp4SdWoAOBlD0NkOm2MVMQyvvd1pQ7yKZ0Qd0fQLMEKNwU7bYhNHgDXhUPCY04Z4DdeLOjXPfBlwkdO2OEAS+DAcCpY7bYiXcLWoI5qeB3wSCDpti8PsBt4Oh4Jxpw3xAq4VdUTTi4GrcX7hxC0cBV4Nh4JNThvidlwp6oimTwcuwUgkUpwkCrweDgUPO22Im3GVqCOaHgCuBCY4bYuLSQLrwqHgVqcNcSuuEXVK0AuBYqdt8QjbwqHgWqeNcCNuyqe+CiXo3jA5ounJcCj4Xo9/IUQmRg5M/05XFkYWYXunqwVoREpf5KK4wlNHNH0OxqKKoveUh0PB9QjRkWI7DBiauvI593FJEmjEGJDqwEGk1C2w13Ic99QRTZ+FEnRa9IvHk5Nqqyc0/XTZ4IEwir79fwYwPgT5wFjgEoRoAfanrgNI6YmND4566oimT8NYWFH0gsLG+ti8yo8TJXXHsjOkNHJfnnu6hXffyLWw2yiwA9iGlK6eVnRM1BFNnwjMc6Rzj1LYWB+7Zs/OZHH98ZzTHkwmJb9YEaNiq9U5MRLYB2x0a3jiiKgjml4I3ICahz4nRjTWt129Z2eiWzF3pqU5zvLF0NhgV1hZAaxHyhab+jsnbBd1RNOzgFtx784U15CZSCQXVWyJTq6tPvewYv+eKE883B8p7XIYcaAc2OSWmNuJXOS5KEGflZGN9bF7NqyJ90rQAMVjs7nhC60WmdUdmcBM4CaEGGJjvz1iq6eOaPp4YL5tHXqUOZW7mi/fvzs3kG54JqXkp8ti7Npud855AliHlNts7vcUbBN1RNNzgNuwvjyBZ8lpa0vctuVv7SObGvouxuN6G4/8Sz+SCSfGLRXAO04t5tgZflyBEnSP5LS1Je7S1iVMETTA4GAW193s1ABuAvCp1Iqm7dgi6oiml2JM6Cu6IbctFr9bW5coiEXNTbOdvyiHocPaTW3z3BkNLEII21OHLRd1al/hJVb341UGxqLxuzeuS+abLWiAzH4B7rjfyY0FI4DrEcLWb2g7PPUYoMCGfjzHwFg0/kXt/WReW8w6bzZ2Qg5jJ0Yta//sDAUWIIRtoa4dHV1sQx+eIyOZSN6+aUPCUkF3cOPtTmetFQFz7OrMUlFHNL0EcMXcpdu4fsfm6JBoiz1fyyXjchgz3ulaImUIMd2Ojqz21CGL2/ckoar9LWX6ESuTj07nxtvdkCt9CUKMsLoTy0SdqtMx3Kr2vUpetLX96j077Z/aLB2fQ1Fxm+39nooA5lk91Welp1ZeugtCSnnLVi3RL5nMcMSAq65zQ4mFfIxldcuwRNQRTQ9i1IdWdGJq9aHWwuZG58qlTZ/Vn8xMN4QhUxGi0KrGrfLU4yxq17MIKeXcfbuc3WnUPzuDi2c7Ob3XgcAoIWcJVolarR52IVR1oHWQHdN3Z2POArdUiR2OEKVWNGz6C4xo+nBgkNnteplAMimv2L+7n9N2ADB6TH9yB7gi7xmjeq3pWPGpLbWgTU8z89C+1gHtbe4QtRCCKRc7PWfdwVCEML0shhWi9npVf9O57MBedwi6g6mWTj70FtMrCZgq6tRBQkPNbNPrFNcdi+XG290l6gsnOR/bn+QChDjz3steYranvsDk9jzPlCNVbpgbPpWc3EyKx7olBAkAF5rdoJlYvgTqNS48WuMmr3iScWVuGSyCcd6laZgt6nyT2/M0o+qPuy/06GD0GKct6EwQIUybMVOitpCpR6rc5A1PZVSxM0v1PWPat7xpok7V87A388zllNQddZtwTjJ0eD+EcDrPujOmpVWY6amVl+6EkFLmxaLuDD3A2Oo1vMip/Yvd4T5PjdqydQoF0Zb4ieKNbmXoMDeFRwVm7WVUntoihjU3uW8qrysFrtuUlGdGI8pTW0RhU4MbUjzPTP5gt9loSjk6M1MhB5jYVo/85PZr7qrZWzE1Kzun8d/eqvhux/1nltw7v+LdN+aLQCA5YvzkzV/55fN/6nhsz4fvDvnV/bcunTTv2pV3Pvb063bYGWxu6lNlpLxvfnlZVmZmNCCEDIhAovqRny7bfuRQ7qInH7unrrV5aEHOgKOr7vvGLycOH5l+wZr8wWnZmAfLsiAaABmARDUsmwJfqUnFxVHIyYbWGnikl027TtS2jPRnLLr1vZy8grdeeeLhf+i49/ZvfzJx/6YNMxav/PB7OXkF8aodm0+Z83x++eLbgiVjt9hhXwf5sdY+fwu+9dVvPT61aPSJAucPPPeb66YXFe94/sv/vOqzTz1x7f3PPX3tm19d8ue0O8grSN82eHwqnLBtC/xnx8/z4NaBkE6RSlNEbWb4YYuo5955/8cFhUXNne9tXPnsvFk3f/GVnLyCOEBR2dTGjsdW/fh7MwYOGV5bMOICW88ezEiaX0p306F90x9acP06gIcWXL+u/GDljD41mNXfdBsTwEaYeT9sSOPprhsoOjbSbz6uF1Zq68YvWzglHLluxr+u//PvSgAaaquzNr703Kc+//2fvWi3TYE+1kYUQvDJny1/sPDbD3zr7v95ci5Ac1tb3qUlF9YDXFpyYX1LW1vfVuFEIC1RC+CT8GAhfOtuozTzCf4dxg+Ahs9ATRpNm+IYzQw/HDsVQCaTgVhzU274lU2RDc//V+nLTyy9d+aNdyx5Zsm9N0xbcOPq/MIi25N3An2sJrvqvm+smF06vl47WDlo4c9XfO3xt16pNsm0k6S5+LIKVsyGeg0GLYSvPQ7VX4ePAf4As+ak56XBJA2Z6V0dG0lnD8w7Pnn+dVogI4NLb7mrUgiSR3ZvH3j0QOWYD1/471uWzh2zbPeGtdfsWPvaomeW3HuVHTYl+lhla3bp+HqA0AWljTOLx5S/u7eidEBWVsMH+3blA3ywb1d+blZW45lbOQtpfpvMBsM2aJwJ5e+mNoa0QOAjuPjB9EVtioZ8IeqSGZeW7/rgnTKAbW+vGp5MJDMLx01qemjVpkeXrtm7ZOmavUvGzZrzRtmchS//3bIn/2qHTck+LEFX1R/P2ndM79/x80dVByZPKxpdNX1Uyablr6+cDbD89ZWzp48q2dQnI9P4NqmCrH2p2LcKsj6CydOgCuBRmBSE6rlQl6ZFpiwGmRl+2LI69aPbrvzysYOVE+Kx6MCHryhZMW3hZ/9ywzeXv/vkl66/65H5Ex4WGRnxq7/y9acDGc6mXbRlZqYt6s2HD+Td+fuf3w+QlDLj8jHjP3j42pu33nr4YOWnf/nYPYMfuueK/OzcY6/ct/jJPhnZ0tRrGzdD3p1g2AYZl8MHD8NWgBdg1lWwvi8W9eG5JzDtJIGIpn8GoxCgAvj0js0tU2uq3J3gteb1Fv70WzfZuAYpt/e1ETPDjwYT2/I8NQM9sKG+/rjTFnTFlENHzRR1vYlteZ6aAYPcncwEUHfUaQu6okTtZmoGDHJv2mkHDXVu+uAlMOnb3swXVWdiW56nNSsrI5aR4abUztOpPuRsGbRTqTXrcFGzY2q3ZX05yvGcXDcl4Z9KS3OChjo3idq0NAbTRB0OBZNA3xYDfMbO4Aj3fshrDrvtA2faiqnZMZWKqzuxpbDIneURAA5Wuik0SgJHzGrMbFG7bo7ISRr7Z2fW5g5wQ+nc09m/27FcnW7Yj5SmnXJgtqgPmdye59kxzKUhyI7NbvoWqTCzMbNFfRhw/948G9lcOMp9U3t6TcxFg8QosN/MBk0VdTgUTGDiKNYPNGTn9KseOMhdIciOj9zkeHYj+5h83gUrJt9N/dT5gTfHljltwql8tMEtRXYkYPo2OytEvRfDWEWK/QVDsqsG5aezZ8982mIJdu2w/8i77tmNlKbPmJku6nAo2IKJ0zN+YfW4MncsSW/VYiQTbpj5SAJ/s6Jhq97o3Ra161mq8gr6H8gb7Ly3fvtVtwwQK6zw0mCdqPdg06YBL7F63MSAdDI0O1rbRuXHbpjKa6VvmwnOiCWiDoeCrcBOK9r2MkcG5ffXRo42ZXdHWqx7yy2zHmuR0rIZISvjvHJUgtNprB5Xlns8O8f+oyni7Unee8MNA8Q9SLnXyg4sE3U4FGwCdlnVvldJBgLijxddLOIiYO8HfsPaVlqanZ7KawHWWt2J1SNyDTW9dxpHBwzMenX8ZPsWZOLtSV5+zmkvHQdetTLs6MBSUYdDwXqMQaOiC5tHjMr9qHBU89n/0gQ2rG2lscHpWY+3kLLWjo7smDvVbOjDk7w8ccqA7cER1g4c3eGl11sdR3fGclGHQ8FjQKXV/XiVFyZPz902zEJhr1nttJcuR8pyOzu0a5XrfVT2Xo/8ZdL03M3Di8wXdv3xNl561tTTZHvJB0hp2Xx0T9gi6nAo2AB8aEdfXuWlsqnmx9jPPJUkHndieV5iFKbpW1m0NLHzBW8mvfKu5w0vT5wy4PVxZS0JIfo+3bdVa2X7pmwTzOotbcBrZlRaShfTyo6dCxFNHwzchLk1/HzH0Oamtlu2anJItCW9AV5Lc5zli3Eglq4FViOloxuwbRU1QETTJwNzbO3UgwSSSblg1/aWUPXB3p2lI6XkFz+MsXOznV5aYsxybTQ74T8dbBc1QETTF5Kqaaw4M2OO6dHrKrYE8tpi55aI9MaLLax8xs6ijweA95HSNZuunRJ1NnAzJh1c43eElHLKkarWOft2ZebHoj2Lu/LjVn70XbtmO45jiPmATf2dM46IGiCi6fnADYCTU06eQkgpLzpSFZ27b1fGaeJuqGtnxUMBmhutzu+oxqhHvdcNoUZ3OCZqgIimB4HPAG7I8fUUk2oOt06rPiSL649lZ7S2SB7/ToKaw1a9j+0YGz+2IqXrSqV2xVFRA0Q0fQSwCDUjkhZZ8XjsjhXfXF/4p9+NBC4AzBog1mNsot4PHHarV+4Ox0UNENH00cCncPDYOo8SB14Jh4Iny1IIMQgY2ukahHFGSzanO4721NWKIeI64ChGBVJTakU7gStEDRDR9HHA1Th4dJ3HSACvhkPBg+f8DCECGKFeHCl9m7bgGlEDRDS9DOOwSSXsM9MGrO6VoM8jXCVqgIimFwPXAO4r1+UO6jA8tKow2wOuEzWcWE6/FiMeVJxkH/BmOBR0W21pV+FKUcOJBZqrMUb0CtgYDgVVpuM54FpRdxDR9IuBT3D+xtntwF/DoaCpO0eEEAmMzMlMjFJxfy+lrBNClALbMUpcZAHvAA8AxZ3ud/C4lPJ3QogfAF8EBkspHV8ldr2oASKaXgTMB3qX3ON9dAxBHzO7YSFEU4cAhRC/BSqklD9IifpFKeUUIUQm8CbwI2Bjx/1u2roMIzT62A2i9sS8cDgUrAKexXhjfTsV1Yl2YB3wf1YIuhvWAaO63pTGtN97wIVnerKU8n0ppWtKOHtmFS8cCsaBDyOavgO4hLO80R5mD7AuHArastNcCJGBMdv0q24ey0099p3UrXFCiPJOf/JPUso1lhvZSzwRfnRHRNMLgcuBYU7bYhJHgPfDoaAtFWM7xdSlGNVHF0opE11iagm8IKVc2jksOUObTW4IPzwr6g4imj4BmIl301h1oDwcCtpaH6VDgEKIfOBF4Dkp5Y97Eq8Stc1END0AlAAXAUUOm3MudGS9bQuHgroTBnQZKIaAF4BxGLG1ErWbiGh6AVCGEXPbuQPkXNAxvtp3Ob2A0lWAQoiVwB+ANfQs6q5Ter9OefcfArdjOJQq4Ckp5VJrX0HP+E7UHUQ0XWB4nfHAaMxLyewtjRhbnnY45ZXPN3wr6q6kdtoUAsOBEcBgrFnQacQoBXEIqErVPFHYyHkj6q5ENL0fhsALgQKMUCUXY3tZT6UJkkAM4+y/1tS/UaAZI7SoDYeC7jpe7jzkvBX1mUiFLpkYmYKZGB49Gg4F7S+Wrug1StQK3+GJZXKFojcoUSt8hxK1wnd4WtRCiKYuv98thPiPLvc2CSH+t8u93wgh9gohyoUQG4UQs1P3PyeE2CqESAohZlr/ChRW4GlRnw0hxCSM13ilEKJrLvZiKeUMIAw8mbq3BaMc2ju2GakwHV+LGmPp9vfAaxglzrrjHVJprFLK7VJKdaipx/FMPnUP5HTJ7x0C/KXT758HFgATgX8ETglDUlyPkYKp8AleF3VrKoQAjJgaIw0VIcQsoFZKuU8IcRD4tRBisDxZcvZRIcS3MQqFf8lesxVW4nVRn4kvAGVCiMrU73nALcBTqd8XSyn/6IRhCmvxZUwtjPJanwOmSSlLpZSlwI0YQlf4HF+KGrgSOCSlPNTp3jvAZCHEyJ6eJIS4KRWqzAZeEkK8arGdCgtQuR8K3+FXT604j1GiVvgOJWqF71CiVvgOJWqF71CiVvgOJWqF71CiVviO/wd/HuqNMCl8KwAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 216x216 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "f, ax = plt.subplots(figsize=(3,3))\n",
    "v = venn2([set(screen_pairs[screen_pairs.hap1_hit]['sorted_gene_pair']), \n",
    "           set(screen_pairs[screen_pairs.rpe1_hit]['sorted_gene_pair'])], ['HAP1','RPE1'], ax=ax, set_colors=['skyblue','red'])\n",
    "for text in v.set_labels: text.set_fontsize(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Merge screen pairs with observed LFC"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Load observed LFC values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_observed_lfc(sheet_name,col_name,new_col_name):\n",
    "    obs_lfc = pd.read_excel(file_data, sheet_name=sheet_name).drop_duplicates() # there are duplicate rows\n",
    "    obs_lfc = obs_lfc.rename(columns={col_name:new_col_name}).reset_index(drop=True)\n",
    "    obs_lfc['sorted_gene_pair'] = obs_lfc.apply(lambda x: '_'.join(sorted([x.gene1, x.gene2])), axis=1)\n",
    "    obs_lfc = obs_lfc[['sorted_gene_pair', new_col_name]]\n",
    "    return obs_lfc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sorted_gene_pair</th>\n",
       "      <th>hap1_early_obs_LFC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABHD12_ABHD12B</td>\n",
       "      <td>-0.067548</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sorted_gene_pair  hap1_early_obs_LFC\n",
       "0   ABHD12_ABHD12B           -0.067548"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hap1_t1 = load_observed_lfc('HAP1.T12','early_mean_observed_LFC','hap1_early_obs_LFC')\n",
    "hap1_t1[:1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sorted_gene_pair</th>\n",
       "      <th>hap1_late_obs_LFC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABHD12_ABHD12B</td>\n",
       "      <td>-0.130963</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sorted_gene_pair  hap1_late_obs_LFC\n",
       "0   ABHD12_ABHD12B          -0.130963"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hap1_t2 = load_observed_lfc('HAP1.T18','late_mean_observed_LFC','hap1_late_obs_LFC')\n",
    "hap1_t2[:1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sorted_gene_pair</th>\n",
       "      <th>rpe1_early_obs_LFC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABHD12_ABHD12B</td>\n",
       "      <td>-0.288012</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sorted_gene_pair  rpe1_early_obs_LFC\n",
       "0   ABHD12_ABHD12B           -0.288012"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rpe1_t1 = load_observed_lfc('RPE1.T18','early_mean_observed_LFC','rpe1_early_obs_LFC')\n",
    "rpe1_t1[:1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sorted_gene_pair</th>\n",
       "      <th>rpe1_late_obs_LFC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABHD12_ABHD12B</td>\n",
       "      <td>-0.300921</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sorted_gene_pair  rpe1_late_obs_LFC\n",
       "0   ABHD12_ABHD12B          -0.300921"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rpe1_t2 = load_observed_lfc('RPE1.T24','late_mean_observed_LFC','rpe1_late_obs_LFC')\n",
    "rpe1_t2[:1]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Merge observed LFC values with screen hits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sorted_gene_pair</th>\n",
       "      <th>hap1_hit</th>\n",
       "      <th>rpe1_hit</th>\n",
       "      <th>n_hits</th>\n",
       "      <th>in_93</th>\n",
       "      <th>in_102</th>\n",
       "      <th>hap1_early_obs_LFC</th>\n",
       "      <th>hap1_late_obs_LFC</th>\n",
       "      <th>rpe1_early_obs_LFC</th>\n",
       "      <th>rpe1_late_obs_LFC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SEC23A_SEC23B</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>-1.896861</td>\n",
       "      <td>-2.079895</td>\n",
       "      <td>-1.644239</td>\n",
       "      <td>-1.96918</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sorted_gene_pair  hap1_hit  rpe1_hit  n_hits  in_93  in_102  \\\n",
       "0    SEC23A_SEC23B      True      True       2   True    True   \n",
       "\n",
       "   hap1_early_obs_LFC  hap1_late_obs_LFC  rpe1_early_obs_LFC  \\\n",
       "0           -1.896861          -2.079895           -1.644239   \n",
       "\n",
       "   rpe1_late_obs_LFC  \n",
       "0           -1.96918  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sorted_gene_pair</th>\n",
       "      <th>hap1_hit</th>\n",
       "      <th>rpe1_hit</th>\n",
       "      <th>n_hits</th>\n",
       "      <th>min_hap1_lfc</th>\n",
       "      <th>min_rpe1_lfc</th>\n",
       "      <th>in_93</th>\n",
       "      <th>in_102</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SEC23A_SEC23B</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>2</td>\n",
       "      <td>-2.079895</td>\n",
       "      <td>-1.96918</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sorted_gene_pair  hap1_hit  rpe1_hit  n_hits  min_hap1_lfc  min_rpe1_lfc  \\\n",
       "0    SEC23A_SEC23B      True      True       2     -2.079895      -1.96918   \n",
       "\n",
       "   in_93  in_102  \n",
       "0   True    True  "
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "screen_pairs_w_lfc = pd.merge(screen_pairs, hap1_t1)\n",
    "screen_pairs_w_lfc = pd.merge(screen_pairs_w_lfc, hap1_t2)\n",
    "screen_pairs_w_lfc = pd.merge(screen_pairs_w_lfc, rpe1_t1)\n",
    "screen_pairs_w_lfc = pd.merge(screen_pairs_w_lfc, rpe1_t2)\n",
    "display(screen_pairs_w_lfc[:1])\n",
    "\n",
    "screen_pairs_w_lfc = screen_pairs_w_lfc.assign(\n",
    "    min_hap1_lfc = screen_pairs_w_lfc.apply(lambda x: min(x.hap1_early_obs_LFC, x.hap1_late_obs_LFC), axis=1),\n",
    "    min_rpe1_lfc = screen_pairs_w_lfc.apply(lambda x: min(x.rpe1_early_obs_LFC, x.rpe1_late_obs_LFC), axis=1))\n",
    "\n",
    "screen_pairs_w_lfc = screen_pairs_w_lfc[['sorted_gene_pair','hap1_hit','rpe1_hit','n_hits','min_hap1_lfc','min_rpe1_lfc',\n",
    "                                         'in_93','in_102']]\n",
    "# Num hits should be the same as seen earlier\n",
    "assert((sum(screen_pairs_w_lfc.hap1_hit)==sum(screen_pairs.hap1_hit)) & \n",
    "       (sum(screen_pairs_w_lfc.rpe1_hit)==sum(screen_pairs.rpe1_hit)))\n",
    "screen_pairs_w_lfc[:1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "screen_pairs_w_lfc.to_csv(file_chymera_pairs_orig)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
